File: tests/python/greedy_algorithms/valid_starting_city/test_valid_starting_city.py
from problems.greedy_algorithms.valid_starting_city.python.valid_starting_city import \
    validStartingCity
from tests.python.utility import load_test_cases


def test_valid_starting_city() -> None:

    test_cases = load_test_cases("greedy_algorithms/valid_starting_city.json")

    for idx, test in enumerate(test_cases):
        distances: list[int] = test["distances"]
        fuel: list[int] = test["fuel"]
        mpg: int = test["mpg"]
        expected: int = test["expected"]

        result: int = validStartingCity(distances, fuel, mpg)

        assert (
            result == expected
        ), f"Test case {idx} failed: expected {expected}, got {result}"

File: tests/cpp/greedy_algorithms/test_valid_starting_city.cpp
#include "../../../problems/greedy_algorithms/valid_starting_city/cpp/valid_starting_city.h"
#include "../jsontestbase.h"

#include <vector>

class ValidStartingCityTest : public JsonTestBase {
public:
  ValidStartingCityTest() {
    json_file_path = std::string(TEST_CASES_DIR) +
                     "/greedy_algorithms/valid_starting_city.json";
  }
};

TEST_F(ValidStartingCityTest, TestCases) {
  ASSERT_FALSE(test_cases.empty()) << "Test cases are empty or not loaded";

  for (const auto &test : test_cases) {
    auto distances = test["distances"].get<std::vector<int>>();
    auto fuel = test["fuel"].get<std::vector<int>>();
    auto mpg = test["mpg"].get<int>();

    auto expected = test["expected"].get<int>();

    auto result = validStartingCity(distances, fuel, mpg);

    EXPECT_EQ(result, expected) << "Failed for test case: " << test.dump(2);
  }
}
